### PR TITLE
Enhancement: Add context provider to sidebar

### LINKED ISSUE #505: Map blocked reasons to pre-filtered cookie table view
## Feature Request

**Description:**
Enhance the user experience by mapping blocked reasons to the pre-filtered cookie table view on the cookies landing page. When users interact with the pie charts or categories displayed below them, clicking on any segment should redirect them to the corresponding pre-filtered cookie table. This functionality ensures that users are presented with a shortlisted view relevant to their selection.

**Motivation:**
By directly linking the pie charts and categories to pre-filtered cookie tables, users can quickly navigate to relevant information without manual filtering, reducing the time and effort required to locate specific data subsets. Additionally, it promotes better utilization of the platform's features, empowering users to gain actionable insights from their cookie data more effectively.

### PR DESCRIPTION
**Pull request by mayan-000**
## Description
This PR adds code to introduce context provider in `useSidebar` hook.

## Relevant Technical Choices
- Removes prop-drilling using the sidebar context selector to extract and update data.
- Updates sidebar type, 
  - Now elements are not stored in state, but callbacks are stored, rendering the JSX.

## Checklist

- [x] I have thoroughly tested this code to the best of my abilities.
- [x] I have reviewed the code myself before requesting a review.
- [x] This code is covered by unit tests to verify that it works as intended.
- [x] The QA of this PR is done by a member of the QA team (to be checked by QA).

Fixes #505

### CODE REVIEW

File: packages/extension/src/view/devtools/components/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    if (Object.keys(tabFrames || {}).includes(currentItemKey || '')) {
+      setSelectedFrame(currentItemKey);
+    } else {
+      setSelectedFrame(null);
+    }
+  }, [currentItemKey, setSelectedFrame, tabFrames]);
+
+  useEffect(() => {
+    (async () => {
+      const tabId = chrome.devtools.inspectedWindow.tabId.toString();
```
Comment (mohdsayed) [inline packages/extension/src/view/devtools/components/layout.tsx:182]: What is the advantage of using `chrome.devtools.inspectedWindow.tabId.toString();` over the utility function `getCurrentTabId`?

File: packages/extension/src/view/devtools/components/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+      const data = await chrome.storage.local.get();
```
Comment (mohdsayed) [inline packages/extension/src/view/devtools/components/layout.tsx:184]: `chrome.storage.local` was found out to be the slowest in Benchmarking Results of Different Chrome Storage, can we change this to session instead while we are refactoring it ? 

https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/403#issuecomment-1884642155

File: packages/extension/src/view/devtools/app.tsx [hunk 3]
```diff
@@ -277,70 +98,23 @@ const App: React.FC = () => {
     })();
   }, [allowedNumberOfTabs]);
 
-  useFrameOverlay(filteredCookies, handleUpdate);
-
   return (
-    <div
-      className="w-full h-screen overflow-hidden bg-white dark:bg-raisin-black"
-      ref={contextInvalidatedRef}
+    <SidebarProvider
```
Comment (mohdsayed) [inline packages/extension/src/view/devtools/app.tsx:102]: ❤️

File: packages/design-system/src/components/sidebar/useSidebar/index.tsx [hunk 4]
```diff
@@ -16,7 +16,15 @@
 /**
```
Comment (amovar18) [inline packages/design-system/src/components/sidebar/useSidebar/index.tsx:16]: Can we split this as well into:
- `context.ts`
- `sidebarProvider.tsx`
- `useSidebar.ts`

**Review by amovar18 — CHANGES_REQUESTED** [review]

File: packages/design-system/src/components/sidebar/useSidebar/index.tsx (hunk 4 above)
Comment (mohdsayed) [reply packages/design-system/src/components/sidebar/useSidebar/index.tsx:16]: @mayan-000 Please resolve the requested change.

File: packages/extension/src/view/devtools/app.tsx [hunk 5]
```diff
@@ -16,309 +16,64 @@
...
-  }, [setContextInvalidated]);
-
-  useEffect(() => {
-    window.addEventListener('mouseover', listenToMouseChange);
-
-    return () => {
-      window.removeEventListener('mouseover', listenToMouseChange);
-    };
-  }, [listenToMouseChange]);
+  const contextInvalidated = useContextInvalidated(contextInvalidatedRef);
```
Comment (mohdsayed) [inline packages/extension/src/view/devtools/app.tsx:38]: Lets move the context invalidated logic to a custom hook and make App look more cleaner. I moved it in the last commit however haven't tested it. Please take a look.

Comment (mayan-000) [reply packages/extension/src/view/devtools/app.tsx:38]: Tested, works as expected 👍

**Review by mohdsayed — APPROVED** [review]

**Review by amovar18 — APPROVED** [review]
LGTM!

**Pull request merged**